keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,770 @@
+   // the same container multiple times, it should achieve the same results.
+   // For example, a quiesce command is run on a database and the database is
+   // quiesced. If the same quiesce command is run again, it should not have
+   // additional impact because the database is already quiesced.
+   Action *ContainerNotifierAction
+ }
+
+// ContainerNotifier describes a command to run in the container.
+type ContainerNotifierAction struct {
+    // handler describes how this notifier is delivered to the container.
+    Handler ContainerNotifierHandler
+
+    // Number of seconds after which the notifier times out.
+    // Defaults to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
+}
+
+// Handler defines a specific action that should be taken
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+}
+```
+
+##### Notification API Object
+
+The Notification object is a request for a ContainerNotifier. It will be an in-tree API object in the "node.k8s.io" API group so that Kubelet can monitor and trigger the actions. The external controller or a user creates the Notification object to request a ContainerNotifier.
+
+All containers in a selected pod which have a "ContainerNotifier" defined in their spec with the "ContainerNotifierName" will have their corresponding "ContainerNotifierAction" executed. Moreover, there is no guarantee of ordering on execution.
+
+```
+type Notification struct {
+           metav1.TypeMeta
+           // +optional
+           metav1.ObjectMeta
+
+           // Spec defines the behavior of a notification.
+           // +optional
+           Spec NotificationSpec
+
+           // Status defines the current state of a notification.
+           // +optional
+           Status NotificationStatus
+}
+
+type NotificationSpec struct {
+           // This contains the name of the ContainerNotifier to send to a container.
+           // Name must be unique within a container.
+           // Names are label-style, with an optional prefix. Names without the prefix
+           // such as `quiesce` are reserved for Kubernetes-defined "well-known" names.
+           // Names with a prefix such as `example.com/label-style` are custom names.
+           ContainerNotifierName string
+
+            // PodSelector specifies a label query over a set of pods.
+            // +optional
+	PodSelector *metav1.LabelSelector
+ }
+```
+
+###### Notification Status
+
+NotificationStatus represents the current state of a Notification. It has a list of PodNotificationStatuses.
+
+A PodNotificationStatus represents the current state of a Notification for a specific pod. A PodNotificationStatus contains ContainerNotificationStatuses of all the containers which have the corresponding ContainerNotifier specified. PodNotificationStatus is created when a Notifier starts to run in a pod.
```

Comment:
This is going to be a problem.  Imagine a notifier that selects 1000 pods.  That's not implausible.  Even 5000 is believable.  That leaves you just a few hundred bytes per pod for status, before it starts failing (too big).  Worse, it will fail in non-deterministic ways in practice.

I think we have a few choices here, and we should seek a pattern precedent (@liggitt for consideration)

1) Notification.status is a "summary".  The real status is stored in another new resource `NotificationResult` which is selected by something like "notification.kubernetes.io/uid=<notification-uid>".  Each NotificationResult holds the status for one single pod-notification.  Need to define the lifecycle management of this so we don't leave them around for too long.

2) Notification.status is a "summary".  The real status is stored in each pod, e.g. in `status.notifications[]` which would be a list of "recent" notifications and the per-container statuses

3) other ideas?

TL;DR: we can't have (practically) unbounded lists in a single resource.

Reply:
Ok, I'll wait for more suggestions.

Reply:
Are there any existing examples that I could follow?

Reply:
EndpointSlice is similar - I am not sure there's a closer example.

Reply:
> NotificationStatus represents the current state of a Notification. It has a list of PodNotificationStatuses.

when we talked about this in doc form, I thought notification was scoped to a single pod in order to allow [notifying in waves](https://docs.google.com/document/d/1SWSlZoxY5zFjBKFKaATP07s3q02UenSp8R9-yRkCcwg/edit?disco=AAAAJ1RQggU) and [avoid unbounded status accumulation](https://docs.google.com/document/d/1SWSlZoxY5zFjBKFKaATP07s3q02UenSp8R9-yRkCcwg/edit?disco=AAAAKEpj35c) ... did that approach get revised since then?

Reply:
So that means we can't have a PodSelector in the NotificationSpec.  It will just be a PodName instead.

Reply:
@thockin would you consider having one to one mapping between a Notification and a Pod?  That would simplify status update.

Reply:
Just read the discussion in the doc, I think the notification per pod makes most sense and seems safer. Also the idea of building on top using a notification controller for safe rollout of notification is even better

Reply:
If we go to 1:1, do we need a higher-level API that selects pods and expands a notification for each of them?or should we add that to kubectl?  or should we just say "someone else's problem"?

If we go 1:1 does a subresource make more sense?  It gives even finer RBAC control, but we need the result to be a new resource (for status).  E.g. POST to /pods/foobar/notify -> creates a PodNotifier; kubelet watches PodNotifiers.  Any precedent for that?  Any need to get finer-grained than per-namespace?

Reply:
I like the subresource idea which i was hinting here https://github.com/kubernetes/enhancements/pull/1995/files#r493239279 . I guess i don't follow why it should result in a new PodNotifier being created? For e.g. does the POST on a log or the exec sub resource result in a new object ? The main reason i like this is its less number of new objects to manager , less strain on etcd, since the notifications are an array. It does make the RBAC more fine grained. Thinking out loud, if i am not mistaken, you can say a service account can or cannot exec to any pods in a namespace, So similar should be possible with this notify subresource